Fix(2788): by preventing invalid route creation

MSE99 opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

#### Note

This PR fixes #2788, when `prefixTrailingSlash` is set to `'both'` fastify will generate two routes, one that has a trailing slash and the prefix and the other without the trailing slash it will also auto generate generate `HEAD` routes (if `exposeHeadRoute` is set to `true`) for both of the new endpoints which was the primary source of the duplication bug, setting `exposeHeadRoute` to `false` when attempting to add the second route seems to fix this issue.

if `prefixTrailingSlash` is set to `'no-slash'` then fastify will create a `GET` route and (if `exposeHeadRoute` is set to `true`) fastify will generate one `HEAD` route however it actually create a route where the prefix is concatenated twice, this is due to the fact that the path passed is an empty string `''` and the options object used to create the `HEAD` route is the "same" one that was used to create the original route, `''` empty strings are not truthy so the `path` string which is found on the `GET` options object is used to create the `HEAD` route, this is fixed by using `/` as a path to create the the `HEAD` route instead of of `''`

mcollina reviewed (approved):
lgtm

L2jLiga reviewed (approved):
LGTM

salmanm reviewed (changes requested):
While this solution fixes the issue reported, I think that it may be a work around which we had to do because of another well hidden bug that's causing this behaviour in the first place.

`prepareRoute ` function sets `url` [here](https://github.com/fastify/fastify/blob/master/lib/route.js#L114). But it doesn't set `path` which is the root cause of the double append route `/prefix/prefix`.

`prepareRoute` is being called with `path` [here](https://github.com/fastify/fastify/blob/master/lib/route.js#L243), but the argument we receive is `url` (not saying that it adds to the problem, just a minor confusion)

If we don't do the 2 changes done in this PR and just add a `opts.method !== 'HEAD'` condition [here](https://github.com/fastify/fastify/blob/master/lib/route.js#L165), it'll fix the issue

I'm requesting changes, but if you guys think the current solution is more appropriate, I'll approve.

MSE99 commented:
@salmanm I just implemented your solution and it fixed the duplication bug (which occurs whenever you set `prefixTrailingSlash` is set to `both` and it tries to create `HEAD` route for the new route with the trailing slash), however it didn't fix the `/prefix/prefix` bug, i think this is due to the fact that the path (`url`) passed to prepareRoute (here)[https://github.com/fastify/fastify/blob/master/lib/route.js#L243] is an empty string, empty strings are not truthy so the `path` of the `GET` route is (used instead)[https://github.com/fastify/fastify/blob/master/lib/route.js#L154] (due to the fact that the options object for the new `HEAD` route is basically created by merging `GET` route object with additional properties so the final path value of the `GET` route is the path value for the `HEAD` route)

salmanm commented:
Can you push your changes?
Did you just rename `url` to `path` in `prepareRoute` or setting them both. It should be latter.

MSE99 commented:
I didn't rename anything (for fear that will break other parts of the route creation), i just removed my changes and only added the `opts.method !== 'HEAD'` as you suggested, i just tried setting the `path` to the `url` in `prepareRoute` and ran all the tests, they all pass and your solution works.

Initially i actually did the same and set the value of the `path` object property to  value of the `url` passed to the function explicitly in the `prepareRoute` function (which fixed the bug that caused `HEAD /prefix/prefix` route to be created), later i removed that and instead passed a '/' path to the `prepareRoute` function only if the path of the `HEAD` route was an empty string.

changing the creation of the opts object for the `HEAD` route to this:
```js
options = Object.assign({}, options, {
      method,
      url,
      path: url,
      handler: handler || (options && options.handler)
 })
```
instead of this:
```js
options = Object.assign({}, options, {
      method,
      url,
      handler: handler || (options && options.handler)
})
```
seems to fix the issue, the `prepareRoute` function is used in other parts of the framework, is this change safe?

MetCoder95 commented:
> I didn't rename anything (for fear that will break other parts of the route creation), i just removed my changes and only added the `opts.method !== 'HEAD'` as you suggested, i just tried setting the `path` to the `url` in `prepareRoute` and ran all the tests, they all pass and your solution works.
> 
> Initially i actually did the same and set the value of the `path` object property to value of the `url` passed to the function explicitly in the `prepareRoute` function (which fixed the bug that caused `HEAD /prefix/prefix` route to be created), later i removed that and instead passed a '/' path to the `prepareRoute` function only if the path of the `HEAD` route was an empty string.
> 
> changing the creation of the options object for the `HEAD` route to this:
> 
> 
> options = Object.assign({}, options, {
>       method,
>       url,
>       path: url,
>       handler: handler || (options && options.handler)
>  })
> instead of this:
> 
> 
> options = Object.assign({}, options, {
>       method,
>       url,
>       handler: handler || (options && options.handler)
> })
> seems to fix the issue, the `prepareRoute` function is used in other parts of the framework, is this change safe?

I would say that yes as it is used to set up a route independently, even tho the function creates a new object every time for each new route that wants to be added. Should introduce a conflict AFAIK.

salmanm commented:
> is this change safe

I believe so. But @mcollina can confirm.

Eomm commented on lib/route.js:
```diff
@@ -112,6 +112,7 @@ function buildRouting (options) {
     options = Object.assign({}, options, {
       method,
       url,
+      path: url,
```
Why this addition?

MSE99 replied:
the options object for the `HEAD` route is created by merging a copy of the option object of the `GET` route with the additional properties listed above, the `url` passed when creating the `HEAD` route is an empty string which is not truthy so the `path` string from the option object is instead used which causes fastify to concatenate the prefix with the already prefixed path of the `GET` route to create the `HEAD` route path.

setting `path` property to the same value as the `url` that was passed fixed the issue.

salmanm reviewed (approved):
lgtm

salmanm commented:
@MSE99 Please rebase to make sure if it still works ok with the change merged in https://github.com/fastify/fastify/pull/2782

MSE99 commented:
Done, i rabsed and fixed all the conflicts.

mcollina reviewed (approved):
lgtm

Eomm commented on test/route.test.js:
```diff
@@ -1034,3 +1034,56 @@ test('Set a custom HEAD route before GET one without disabling exposeHeadRoutes
+  const trailingSlashReply = await fastify.inject({ url: '/prefix/', method: 'HEAD' })
+  const noneTrailingReply = await fastify.inject({ url: '/prefix', method: 'HEAD' })
+
+  t.ok(doublePrefixReply.statusCode === 404)
+  t.ok(trailingSlashReply.statusCode === 200)
+  t.ok(noneTrailingReply.statusCode === 200)
```
In case of errors the `tap` message would be clearer in this way

```suggestion
  t.equals(doublePrefixReply.statusCode, 404)
  t.equals(trailingSlashReply.statusCode, 200)
  t.equals(noneTrailingReply.statusCode, 200)
```

delvedor reviewed (approved):
LGTM
